router/core/transport.go

```diff
@@ -189,68 +190,42 @@ func (ct *CustomTransport) allowSingleFlight(req *http.Request) bool {
 	return true
 }
 
-func (ct *CustomTransport) roundTripSingleFlight(req *http.Request) (*http.Response, error) {
-
-	var (
-		buf *bytes.Buffer
-	)
-
-	if req.ContentLength > 0 {
-		buf = bytes.NewBuffer(make([]byte, 0, req.ContentLength))
-	} else {
-		buf = bytes.NewBuffer(make([]byte, 0, 1024))
-	}
-
-	keyGen := pool.Hash64.Get()
-	defer pool.Hash64.Put(keyGen)
-
-	// Hash the request body
-	if req.Body != nil {
-		_, err := buf.ReadFrom(req.Body)
-		if err != nil {
-			return nil, err
-		}
-		body := buf.Bytes()
-		_, err = keyGen.Write(body)
-		if err != nil {
-			return nil, err
-		}
-		// Restore the body
-		req.Body = io.NopCloser(buf)
+var (
+	ctBufPool = &sync.Pool{
+		New: func() any {
+			return &bytes.Buffer{}
+		},
 	}
+)
 
-	unsortedHeaders := make([]string, 0, len(req.Header))
-
-	for key := range req.Header {
-		value := req.Header.Get(key)
-		unsortedHeaders = append(unsortedHeaders, key+value)
-	}
+func getBuffer() *bytes.Buffer {
+	return ctBufPool.Get().(*bytes.Buffer)
+}
 
-	sort.Strings(unsortedHeaders)
-	for i := range unsortedHeaders {
-		_, err := keyGen.Write(unsafebytes.StringToBytes(unsortedHeaders[i]))
-		if err != nil {
-			return nil, err
-		}
-	}
+func releaseBuffer(buf *bytes.Buffer) {
+	buf.Reset()
+	ctBufPool.Put(buf)
+}
 
-	sum := keyGen.Sum64()
-	key := strconv.FormatUint(sum, 10)
+func (ct *CustomTransport) roundTripSingleFlight(req *http.Request) (*http.Response, error) {
 
 	// We need to use the single flight group to ensure that the request is only sent once
-	v, err, shared := ct.sf.Do(key, func() (interface{}, error) {
+	v, err, shared := ct.sf.Do(ct.singleFlightKey(req), func() (interface{}, error) {
 		res, err := ct.roundTripper.RoundTrip(req)
 		if err != nil {
 			return nil, err
 		}
-		buf.Reset()
+		buf := getBuffer()
+		defer releaseBuffer(buf)
 		_, err = buf.ReadFrom(res.Body)
```

Comment:
Luckily, file uploads are mutations and we don't allow sf. Could you add a comment about this here?